xds/src/main/java/io/grpc/xds/EdsLoadBalancer.java

```diff
@@ -0,0 +1,737 @@
+        new LoadStatsStoreImpl(), OrcaPerRequestUtil.getInstance(), OrcaOobUtil.getInstance());
+  }
+
+  @VisibleForTesting
+  EdsLoadBalancer(
+      Helper helper,
+      EndpointWatcherImpl.PickerFactory pickerFactory,
+      LoadBalancerRegistry lbRegistry,
+      ThreadSafeRandom random,
+      LoadStatsStore loadStatsStore,
+      OrcaPerRequestUtil orcaPerRequestUtil,
+      OrcaOobUtil orcaOobUtil) {
+    this.helper = checkNotNull(helper, "helper");
+    this.channelLogger = helper.getChannelLogger();
+    // TODO(zdapeng): TBD: handle enableLrs from ClusterUpdate is false/changed.
+    this.loadStatsStore = checkNotNull(loadStatsStore, "loadStatsStore");
+
+    // The following fields are injected for testing
+    this.pickerFactory = checkNotNull(pickerFactory, "pickerFactory");
+    this.lbRegistry = checkNotNull(lbRegistry, "lbRegistry");
+    this.random = checkNotNull(random, "random");
+    this.orcaPerRequestUtil = checkNotNull(orcaPerRequestUtil, "orcaPerRequestUtil");
+    this.orcaOobUtil = checkNotNull(orcaOobUtil, "orcaOobUtil");
+  }
+
+  @Override
+  public void handleResolvedAddresses(ResolvedAddresses resolvedAddresses) {
+    channelLogger.log(ChannelLogLevel.DEBUG, "Received ResolvedAddresses '%s'", resolvedAddresses);
+
+    AtomicReference<XdsClient> xdsClientRef = resolvedAddresses.getAttributes()
+            .get(XdsAttributes.XDS_CLIENT_REF);
+    if (xdsClientRef == null) {
+      // TODO(zdapeng): create a new xdsClient if no one exists.
+      throw new UnsupportedOperationException(
+          "XDS_CLIENT_REF attributes not available in resolvedAddresses " + resolvedAddresses);
+    }
+
+    if (xdsClientRef.get() == null) {
+      // The XdsClient has been shutdown. The balancer will soon shutdown.
+      return;
+    }
+
+    if (xdsClient != null) {
+      Preconditions.checkState(xdsClient == xdsClientRef.get(), "The XdsClient is changed");
+    } else {
+      xdsClient = xdsClientRef.get();
+    }
+
+    Object lbConfig = resolvedAddresses.getLoadBalancingPolicyConfig();
+    checkArgument(
+        lbConfig != null && EdsConfig.class.isAssignableFrom(lbConfig.getClass()),
+        "Expecting an EDS LB config, but the actual LB config is '%s'",
+        lbConfig);
+
+    EdsConfig edsConfig = (EdsConfig) lbConfig;
+    String edsServiceName = edsConfig.name;
+
+    // TODO(zdapeng): If localityPickingPolicy is changed, swap to new policy.
+    // Right now we have only one default localityPickingPolicy (hardcoded RoundRobin),
+    // so ignoring localityPickingPolicy for now.
+
+    if (!edsServiceName.equals(this.edsServiceName)) {
+      if (endpointWatcher != null) {
+        // TODO(zdapeng): Maybe gracefully swap until the localities for the new watcher is READY?
+        xdsClient.cancelEndpointDataWatch(this.edsServiceName, endpointWatcher);
+        endpointWatcher.reset();
+      }
+      endpointWatcher = new EndpointWatcherImpl(
```

Comment:
Why put everything into `EndpointWatcherImpl`'s implementation and then delegate `EdsLoadBalancer`'s work to it? `EndpointWatcherImpl` should just be something to receive `EndpointUpdate`, computations for load balancing  should be in `EdsLoadBalancer`.